### DISCUSSION TITLE
npf_(v)snprintf return negative values

### DISCUSSION THREAD

**Discussion by Gozem** [charlesnicholson/nanoprintf#248] [Q&A] [answered]
Is it possible for npf_(v)snprintf to ever return a negative value. Indicating some kind of internal error. I've tried to read the code and can't find any path where a negative value is returned. 

npf_ftoa_rev() returns negative values, but that seems only to be a trick to mark if inf_or_nan was found and the return value is later inverted to positive again.

A topic of the return value in README.md would be good. Much like man-pages usually have a RETURN VALUE section.

**Comment by charlesnicholson** [comment]
Hi @Gozem - nanoprintf is intended to be a drop-in replacement for (v)snprintf, so it matches the return codes of the standard library functions.

Specifically, from https://en.cppreference.com/w/c/io/fprintf:
<img width="803" alt="Screenshot 2023-09-13 at 1 18 18 PM" src="https://github.com/charlesnicholson/nanoprintf/assets/3010295/149a2085-dd30-479a-b071-911e1115ecbe">

Section 4 above describes the return values that snprintf can return. (vsnprintf is similar but found over here: https://en.cppreference.com/w/c/io/vfprintf).

Because nanoprintf does not deal with `wchar_t` and ignores the `l` length modifier on strings, and instead only deals with the ASCII character set (or the first 128 code points of UTF-8 if you prefer), it will never encounter what is described as an "encoding error". It will only ever return what the text above specifies: the number of characters that would have been written to your buffer if the buffer's size was ignored.

This is good feedback, though, so thanks! I'll add an issue to the tracker to make this explicit in the README.md, and also in the header file itself.

**Comment by charlesnicholson** [comment]
Filed https://github.com/charlesnicholson/nanoprintf/issues/249 for the docs. It should be quick + small; I'll try to get to it soon!

**Comment by charlesnicholson** [comment, marked as answer]
Readme and .h file both now discuss the return values of the `npf_` functions. https://github.com/charlesnicholson/nanoprintf/issues/249

**Reply by Gozem** [reply]
Thank you! For everything, and the fast reply.

I am very impressed with the performance (size and functions) of nanoprintf. It is an absolute true match for any STM32 project.
